src/plugindata.h

```diff
@@ -263,8 +265,66 @@ void plugin_configure_single(GtkWidget *parent);
 void plugin_help(void);
 void plugin_cleanup(void);
 
+/** Called by Geany when a plugin library is loaded.
+ *
+ * This is the original entry point. Implement and export this function to be loadable at all.
+ * Then fill in GeanyPlugin::info and GeanyPlugin::funcs of the passed @p plugin. Finally
+ * GEANY_PLUGIN_REGISTER() and specify a minimum supported API version.
+ *
+ * For all glory details please read @ref howto.
+ *
+ * @param plugin The unique plugin handle to your plugin. You must set some fields here.
+ * @param module The GModule associated with your plugin library.
+ * @param geany_api_version The API version exposed by the Geany instance that loads your plugin.
+ *           This probably not the same as the one your plugin was compiled against.
+ *
+ * @since 1.26 (API 225)
+ * @see @ref howto
+ */
+void geany_load_module(GeanyPlugin *plugin, GModule *module, gint geany_api_version);
+
 #endif
 
+/** Callback functions that need to be implemented for every plugin.
+ *
+ * These callbacks should be registered by the plugin within Geany's call to
+ * geany_load_module() by calling geany_plugin_register() with an instance of this type.
+ *
+ * Geany will then call the callbacks at appropriate times. Each gets passed the
+ * plugin-defined data pointer as well as the corresponding GeanyPlugin instance
+ * pointer.
+ *
+ * @since 1.26 (API 225)
+ * @see @ref howto
+ **/
+struct GeanyPluginFuncs
+{
+	/** Array of plugin-provided signal handlers @see PluginCallback */
+	PluginCallback *callbacks;
+	/** Called to initialize the plugin, when the user activates it (must not be @c NULL) */
+	gboolean    (*init)      (GeanyPlugin *plugin, gpointer pdata);
+	/** plugins configure dialog, optional (can be @c NULL) */
+	GtkWidget*  (*configure) (GeanyPlugin *plugin, GtkDialog *dialog, gpointer pdata);
+	/** Called when the plugin should show some help, optional (can be @c NULL) */
+	void        (*help)      (GeanyPlugin *plugin, gpointer pdata);
+	/** Called when the plugin is disabled or when Geany exits (must not be @c NULL) */
```

Comment:
About `set_data()`, this discussion is showing a problem IMO (that currently isn't, but will be if we want to stop calling load_module() more than once per Geany run): the moment when the cleanup function should be called changes depending from where `set_data()` was called.

* If you call it in `load_module()`, the cleanup should be called in its "counterpart", e.g. when the plugin is unloaded (that would match module unload, unless `make_resident()` was used).
This is because the data created in `load_module()` has to live throughout the lifetime of the plugin -- e.g. survive `init()`/`cleanup()` pairs.
* If however you call it in `init()`, it should be called right after `cleanup()` to do the expected thing (at least what I would expect).
This is because the data was created in `init()`, and so might not be valid (or at least, useful) after the plugin is deactivated.  Moreover, if `set_data()` can only be called once, it has to do something so next `init()` will still work.

Currently, it's the same as we always unload the module when deactivating it, but whether doing so makes sense is questionable.  But if we don't, we then need to know when `set_data()` was called to be able to cleanup at the right time.  This might be tricky, but could work with e.g. a reference count, like this:

```C
on_load_module() = on_init() = {
  if (plugin->user_data_free_func)
    plugin->user_data_ref_count++;
}

on_unload() = on_cleanup() = {
  if (plugin->user_data_ref_count > 0 && --plugin->user_data_ref_count == 0)
    plugin->user_data_free_func(plugin->user_data);
}
```

---

Also, as @codebrainz started using the `free_func` in an incorrect manner, it might mean the docs needs to be more clear -- if @codebrainz got it wrong, I would think other will.  @codebrainz, have you an idea how to formulate this so it'd be clear?